### PR TITLE
iphlpapi support

### PR DESCRIPTION
**Pull request by hankbao**
It looks like the #772 has been stalled for a while. So I'd like to give it a try.

### CODE REVIEW

**Comment by hankbao** [comment]
Added SIZEOF_* functions

**Comment by hankbao** [comment]
@retep998 Hi, I've fixed a bunch of issues and added necessary tests. Now all tests passed. Please take some time to review this PR. Thanks.

File: src/shared/ipmib.rs
Comment (retep998) [inline src/shared/ipmib.rs]: This is UB because it dereferences a null pointer.

File: src/shared/ntddndis.rs
Comment (retep998) [inline src/shared/ntddndis.rs]: Remove these comments

File: src/shared/tcpmib.rs
Comment (retep998) [inline src/shared/tcpmib.rs]: See above regarding UB.

File: src/shared/tcpmib.rs
Comment (retep998) [inline src/shared/tcpmib.rs]: Remove these comments.

File: src/shared/udpmib.rs
Comment (retep998) [inline src/shared/udpmib.rs]: See above regarding UB.

File: src/um/ipexport.rs
Comment (retep998) [inline src/um/ipexport.rs:32]: `__ptr32` is so weird...

File: src/um/iphlpapi.rs
Comment (retep998) [inline src/um/iphlpapi.rs]: This should be defined with `DECLARE_HANDLE!`

**Review by retep998 — CHANGES_REQUESTED** [review]

File: src/shared/ipmib.rs
Comment (hankbao) [reply src/shared/ipmib.rs]: I'm not sure if I understand what UB means correctly. But this is actually doing pointer arithmetic and give out the offset of `table` and I believe it [runs well](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=fce1a2aad45bd8b8003c467342bd50b5)

Could you please elaborate on the matter?

File: src/um/ipexport.rs
Comment (hankbao) [reply src/um/ipexport.rs:32]: Yea, I can't figure out how to do it in rust. So I just leave a `u32` here.

File: src/shared/ipmib.rs
Comment (retep998) [reply src/shared/ipmib.rs]: UB is Undefined Behavior. Dereferencing a null pointer is undefined behavior. Just because it currently happens to do the right thing does not mean it is guaranteed to behave correctly and it could cause horrible things to happen at any time depending on LLVM optimizations.

Comment (hankbao) [reply src/shared/ipmib.rs]: If you're referencing the [rust reference](https://doc.rust-lang.org/reference/behavior-considered-undefined.html), then yes, dereferences a null pointer is listed on it.

Do you think it's acceptable if I do something like [this](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2ff6cbe8b43643ec91b45dfae2e8799c)

Comment (MSxDOS) [reply src/shared/ipmib.rs]: There's currently no way in Rust to take an offset of a field without UB, so people (including myself) who implement their own ways of doing it are doing so at their own risk.
See https://github.com/Gilnaa/memoffset/issues/24 and https://github.com/rust-lang/rfcs/pull/2582

Comment (hankbao) [reply src/shared/ipmib.rs]: > There's currently no way in Rust to take an offset of a field without UB, so people (including myself) who implement their own ways of doing it are doing so at their own risk.
> See [Gilnaa/memoffset#24](https://github.com/Gilnaa/memoffset/issues/24) and [rust-lang/rfcs#2582](https://github.com/rust-lang/rfcs/pull/2582)

Thanks for pointing out the `offset_of` discussion.

But is it UB even with doing pointer arithmetic of a `mem::zeroed() T` in this particular scenario like: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2ff6cbe8b43643ec91b45dfae2e8799c ?

Comment (hankbao) [reply src/shared/ipmib.rs]: If there is no defined way to do `offset_of`, maybe I could replace all these `SIZEOF_*` 
 with some `FIXME` comments as @retep998 mentioned in #676?

Comment (hankbao) [reply src/shared/ipmib.rs]: I've replaced all `SIZEOF_*` in this PR with `FIXME` comments.

File: src/um/ipexport.rs
Comment (hankbao) [reply src/um/ipexport.rs:32]: Since there is nothing we can do about `__ptr32`, mark this as resolved.

**Comment by hankbao** [comment]
@retep998 Please continue your review. Thanks.

File: src/shared/ipmib.rs
Comment (MSxDOS) [reply src/shared/ipmib.rs]: > But is it UB even with doing pointer arithmetic of a `mem::zeroed() T` in this particular scenario like:

Should be fine for winapi structs as `zeroed` is a valid state there.

Comment (retep998) [reply src/shared/ipmib.rs]: I'd still rather they just remain commented out for now and later I'll figure out what I want to do with them.

Comment (hankbao) [reply src/shared/ipmib.rs]: > > But is it UB even with doing pointer arithmetic of a `mem::zeroed() T` in this particular scenario like:
> 
> Should be fine for winapi structs as `zeroed` is a valid state there.

I guess so. I've commented out these `SIZEOF_*`s anyway.

> I'd still rather they just remain commented out for now and later I'll figure out what I want to do with them.

Sure.

**Comment by TyPR124** [comment]
Is there anything here that still needs changed/fixed here, or just waiting on review?

Also I am not seeing the [ICMP API](https://docs.microsoft.com/en-us/windows/win32/api/icmpapi/nf-icmpapi-icmpsendecho) in here. Unless someone else plans to or has already done this, I can work on that once this is merged.

**Comment by hankbao** [comment]
Still waiting for this PR to get reviewed.

File: src/um/ipexport.rs
Comment (TyPR124) [inline src/um/ipexport.rs:14]: I wanted to point out that I think this definition of IPAddr may cause issues. This allows one to write 

```rust
let rust_ip = Ipv4Addr::new(1,2,3,4);
let win_ip: IPAddr = rust_ip.into();
```

However this actually swaps the byteorder. A packet sent to the `win_ip` value will go to 4.3.2.1 instead of 1.2.3.4.

Comment (retep998) [reply src/um/ipexport.rs:14]: Regardless, it is the correct definition, and you just have to be really careful about byte order when working with windows API.

Comment (TyPR124) [reply src/um/ipexport.rs:14]: I don't disagree that it is "correct", but perhaps it is worth adding a small doc comment or something?

I was (until just now) actually under the impression that IPAddr should have been an alias for IN_ADDR, considering everywhere I see IPAddr in MS documentation, it links to IN_ADDR instead. Looking at the header file, I see I was incorrect, though docs do say that the types are compatible.

I'm just concerned that this is a very easy mistake to make outside of an unsafe block and there is no documentation that states this is a problem.

Comment (hankbao) [reply src/um/ipexport.rs:14]: > ```rust
> let rust_ip = Ipv4Addr::new(1,2,3,4);
> let win_ip: IPAddr = rust_ip.into();
> ```
> 
> However this actually swaps the byteorder. A packet sent to the `win_ip` value will go to 4.3.2.1 instead of 1.2.3.4.

> I was (until just now) actually under the impression that IPAddr should have been an alias for IN_ADDR, considering everywhere I see IPAddr in MS documentation, it links to IN_ADDR instead. Looking at the header file, I see I was incorrect, though docs do say that the types are compatible.

This issue has nothing to do with MSDN docs. You may check the rust docs and find out things like this:

```rust
impl From<Ipv4Addr> for u32
fn from(ip: Ipv4Addr) -> u32
Converts an Ipv4Addr into a host byte order u32.
```

`rust_ip.into()` produces a host byte order u32 and you're likely working on a little-endian machine.

Comment (TyPR124) [reply src/um/ipexport.rs:14]: I'll accept that nothing will be done to address this; so be it. I still feel very uncomfortable about the fact that

```rust
let ip = IPAddr::from(Ipv4Addr::new(...));
```

looks like completely correct code, but it is not. It's not obvious that u32 is even involved. Nor does any bit of documentation make this obvious. The *only* way you know this is invalid is to actually look at the definition of IPAddr and see that it is actually u32.

> you're likely working on a little-endian machine.

As far as I know, Windows only runs on little endian so this is a potential issue for every user, which is why I even thought bring this up. I still think a small doc comment would go a long way, but again I accept that this won't happen.

**Comment by ErichDonGubler** [comment]
@hankbao @retep998: Anything stopping this from getting merged? I'd really like to see https://github.com/heim-rs/heim/issues/29 resolved and am not sure how to help!

**Comment by hankbao** [comment]
Need to get reviewed by @retep998

**Comment by GuillaumeGomez** [comment]
@retep998 Need a hand?

**Comment by retep998** [comment]
@GuillaumeGomez I've been far too busy to give `winapi` the attention it deserves, so I'd be happy to have some help.

**Comment by GuillaumeGomez** [comment]
Ok, let's talk about it on IRC whenever we're both present at the same time!

**Comment by vkrasnov** [comment]
Hi all, this would be very useful to me. I can help reviewing if that helps.

**Comment by yuguorui** [comment]
Please...

File: src/shared/ifdef.rs
Comment (retep998) [inline src/shared/ifdef.rs]:
```suggestion
pub const NET_IF_COMPARTMENT_SCOPE_ALL: NET_IF_COMPARTMENT_SCOPE = -1i32 as u32;
```
As per https://github.com/retep998/winapi-rs/blob/0.3/CONTRIBUTING.md#constants

File: src/shared/ifdef.rs
Comment (retep998) [inline src/shared/ifdef.rs]:
```suggestion
    Length: USHORT,
```

File: src/shared/ifdef.rs
Comment (retep998) [inline src/shared/ifdef.rs]:
```suggestion
ENUM!{enum IF_OPER_STATUS {
    IfOperStatusUp = 1,
    IfOperStatusDown,
    IfOperStatusTesting,
    IfOperStatusUnknown,
    IfOperStatusDormant,
    IfOperStatusNotPresent,
    IfOperStatusLowerLayerDown,
}}
```
Enum definitions should match the headers, and in the header only the value for the first variant is specified.
Check the other enums to make sure they match as well.

File: src/shared/netioapi.rs
Comment (retep998) [inline src/shared/netioapi.rs]:
```suggestion
    Type: ULONG,
    TunnelType: TUNNEL_TYPE,
```

File: src/shared/netioapi.rs
Comment (retep998) [inline src/shared/netioapi.rs]:
```suggestion
BITFIELD!{MIB_IF_ROW2_InterfaceAndOperStatusFlags bitfield: BOOLEAN [
```
Make sure to match the types the headers use.

File: src/shared/nldef.rs
Comment (retep998) [inline src/shared/nldef.rs]:
```suggestion
pub const NET_IF_CURRENT_SESSION: ULONG = -1i32 as u32;
```

**Review by retep998 — CHANGES_REQUESTED** [review]
Sorry for taking so long to get to this.

File: src/shared/netioapi.rs
Comment (retep998) [inline src/shared/netioapi.rs]:
```suggestion
    Type: IFTYPE,
```

File: src/shared/nldef.rs
Comment (retep998) [inline src/shared/nldef.rs]:
```suggestion
ENUM!{enum NL_SUFFIX_ORIGIN {
    NlsoOther = 0,
    NlsoManual,
    NlsoWellKnown,
    NlsoDhcp,
    NlsoLinkLayerAddress,
    NlsoRandom,
    IpSuffixOriginOther = 0,
    IpSuffixOriginManual,
    IpSuffixOriginWellKnown,
    IpSuffixOriginDhcp,
    IpSuffixOriginLinkLayerAddress,
    IpSuffixOriginRandom,
    IpSuffixOriginUnchanged = 1 << 4,
}}
```
Make sure to match what the headers do.

File: src/um/iptypes.rs
Comment (retep998) [inline src/um/iptypes.rs]: As this is an *anonymous* union of a single field and a corresponding bitfield, this should follow https://github.com/retep998/winapi-rs/blob/0.3/CONTRIBUTING.md#union-with-a-primitive-field-and-an-anonymous-bitfield-struct-of-the-same-type and have the bitfield applied directly to a `Flags` field in `IP_ADAPTER_ADDRESSES_LH` without the intermediary `IP_ADAPTER_ADDRESSES_LH_s`.

**Comment by hankbao** [comment]
@retep998 all issues resolved. Please continue your review. Hopefully, this PR will get merged shortly.

**Pull request merged**
